daemon: decompose access check into smaller primitive access checkers

jhenstridge opened the pull request:
This is an attempt to make the access checks for snapd's REST API more modular.  As a first step, I've created a number of small primitive access check routines that will either grant access, refuse access, or defer to the next checker in the list.

The `Command.canAccess` method creates a list based on the command's `RootOnly`, `SnapOK`, `GuestOK`, `UserOK`, and `PolkitOK` fields and then runs through the list.

pedronis commented:
@jhenstridge first note:  this two code paths are not tested on master, they should grow tests before we change the code:

* https://github.com/snapcore/snapd/blob/master/daemon/daemon.go#L150
* https://github.com/snapcore/snapd/blob/master/daemon/daemon.go#L159

pedronis commented on daemon/access.go:
```diff
@@ -0,0 +1,149 @@
+
+func (c denySnapSocket) canAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessUnauthorized
```
everything else either returns OK or Unknown, but this one does this... I'll expand on this in the next comment

pedronis commented on daemon/daemon.go:
```diff
@@ -129,85 +116,50 @@ var polkitCheckAuthorization = polkit.CheckAuthorization
-			return accessCancelled
-		} else {
-			logger.Noticef("polkit error: %s", err)
+	for _, rule := range c.accessRules() {
```
because of the case where we get an  accessUnauthorized  this is still order dependent, it makes it more fragile than the current code. If we want to go in this direction I think we need to make the order of the rules irrelevant for that this should work more like:

* if some rules return OK and all the other returns unknown return OK
* if a rule returns Unauhorized return immediately with that
* if all rules returned Unknown return Unauhorized as well

jhenstridge replied:
Ordering is important for the polkit rule, as it potentially has the side effect of prompting the user for their password.  We only want to do that as a last resort if no other checkers grant access.

Based on last week's conversations where you said it would be better to make the checks explicitly procedural if ordering is important.  Looking at the existing REST endpoints, I think these could be split into three separate checkers:

1. Optionally allow GET requests from everyone, and require one of macaroon, root, or polkit for other methods.  Requests from `snapd-snap.socket` are denied.
2. Something equivalent to the current `RootOnly` behaviour.
3. Something equivalent to the current `SnapOK` behaviour.

The checker for the theme API could slot in next to these three.

pedronis replied:
That proposal is not too dissimilar to how we describe/classify Access here:

https://forum.snapcraft.io/t/snapd-rest-api/17954

as either  open, root, authenticated.   SnapOK is a bit orthogonal to the rest so we need to be careful.

In case we are looking for defaults (not sure):

* GET should default to authenticated
* everything else if we go with a default should be root-only

jhenstridge replied:
I opted not to have a default access policy, and instead added a test to make sure every command has a non-nil access checker corresponding to the HTTP methods it implements.

jhenstridge commented:
As far as the question of Unauthorized vs. Forbidden, the standard pattern from HTTP is that Unauthorized should be returned if the request could be satisfied with an appropriate `Authorization` request header.  Forbidden should be returned if the request has an `Authorization` header and it's still not allowed, or if adding the header wouldn't help.

This isn't handled particularly well in the current version of this branch.  For the RootOnly and SnapOK cases, access failures should be Forbidden.  For the cases where macaroon auth is allowable, it should return Unauthorized if no macaroon is provided.

jhenstridge commented:
I've updated the initial comment to describe the new approach.  Most of the increased size is down to updating each `Command` struct.

While going through the set of commands, I found one with a nonsensical set of flags: `/v2/snaps/{name}/file`.  It had `UserOK` and `PolkitOK` set, but only defined a GET method.  The Polkit action would never be checked since GET requests are allowed without authentication and there are no other methods.

So I've added a test to ensure we're only setting the new `ReadAccess` if there is a GET method, and only setting `WriteAccess` if there is either a PUT or POST method.  This ensures we can't make access decisions about API that doesn't yet exist.

jhenstridge commented:
One other possible simplification would be to get rid of the `accessResult` type, and have `checkAccess` return a `Response`.  If the checker returns nil, it would be treated as `accessOK` is today.  If it returns non-nil, that response is sent to the client and the request is considered complete.

This PR has grown fairly large though, so perhaps that would be better off in a follow up.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+type openAccess struct{}
```
Please document that `openAccess` is not free-for-all and that it filters-out access over the snap-specific socket.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+
+func (ac openAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Forbid access from snapd-snap.socket
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
```
Can `ucred` ever be legitimately nil? I'm worried that this should be written like this instead:
```
if ucred == nil {
    return accessForbidden
}
if ucred.socket == dirs.SnapSocket {
    return accessForbidden
}
return accessOK
```

```

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	if authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err == nil {
```
Question: what kind of details could we provide? (the `nil` argument here)

jhenstridge replied:
The argument is described here in the documentation:

https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.Authority.html#eggdbus-method-org.freedesktop.PolicyKit1.Authority.CheckAuthorization

There are two main uses:

1. override the icon and/or message registered for the chosen action.
2. provide values to expand `$(key)` substitutions in the message.

We are not currently using either of these features, hence the `nil`.  This could be useful for e.g. exposing the snap name in a prompt asking for permission to install a snap.  Doing so would probably require further changes to how we handle authentication, especially if the data we want to substitute is held in the request body.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	if authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err == nil {
```
Stylistic nitpick: I would use a switch err with cases nil, polkit.ErrDismissed and default for the logging clause.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+		return accessOK
+	}
+
+	if ac.Polkit != "" {
```
Could you add a note that polkit is probed last to avoid interactive prompts. I know you know this but it might help others reading the code later.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+type rootAccess struct{}
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.uid == 0 && ucred.socket != dirs.SnapSocket {
```
Again, same request. If `ucred == nil` we should fail immediately. Having done that we can check for the socket type and then finally `ucred.uid == 0`. This will effectively state the pre-conditions very strongly and not combine them into one logical expression.

zyga commented on daemon/access.go:
```diff
@@ -0,0 +1,143 @@
+type snapAccess struct{}
+
+func (ac snapAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
```
Once again please separate the `ucred != nil` check.

On some systems snaps *may* access the system socket so perhaps we should allow both? CC @jdstrand

jhenstridge replied:
It looks like the Go client library and snapd-glib will both use `snapd.socket` to call the snapctl REST API unless configured not to.  So I guess it is possible that there are some snaps with `snapd-control` that could be broken by this.  From the look of it, the existing code will let the call succeed if the snap is running as root.

So perhaps it would be best to remove the socket check all together from this checker.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	// Access allowed from other sockets, or without peer credentials
+	ucred.socket = dirs.SnapdSocket
+	c.Check(ac.checkAccess(nil, ucred, nil), Equals, accessOK)
+	c.Check(ac.checkAccess(nil, nil, nil), Equals, accessOK)
```
IMO this test case should change, for the reasons described above.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	c.Check(ac.checkAccess(req, ucred, user), Equals, accessForbidden)
+
+	// With macaroon auth, a normal user is granted access
+	ucred = &ucrednet{uid: 42, pid: 100}
```
Nitpick, should we explicitly check that socket is set to the system socket? An empty socket feels wrong to allow.

jhenstridge replied:
Perhaps.  Based on your comment about `snapAccess`, maybe the other checkers should explicitly check that the request is received on `snapd.socket` and `snapAccess` allow any socket.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	c.Check(ac.checkAccess(req, ucred, user), Equals, accessOK)
+
+	// Macaroon access is also possible without peer credentials
+	c.Check(ac.checkAccess(req, nil, user), Equals, accessOK)
```
Can you give an example of such access? Would this be the future http-only access or is that something that we may expect to see now?

jhenstridge replied:
This is the future case of API over TCP, yeah.  As discussed earlier, I agree that it makes sense to have this blocked by default so that we will need to actively decide how authentication should be handled at that point.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	c.Check(ac.checkAccess(req, ucred, nil), Equals, accessUnauthorized)
+
+	// The root user is granted access without a macaroon
+	ucred = &ucrednet{uid: 0, pid: 100}
```
Same remark about not having a socket.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (s *accessSuite) TestAuthenticatedAccessPolkit(c *C) {
+	defer func() {
```
This feels like a missing `MockCheckPolkitAuthorization` that is hand-rolled inline. I'd prefer a more explicit approach.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	var ac accessChecker = authenticatedAccess{Polkit: "action-id"}
+
+	var logbuf bytes.Buffer
+	log, err := logger.New(&logbuf, logger.DefaultFlags)
```
We have `MockLogger` for this, please use it instead. This will also remove the need for half of the special-cased local mocking I commented on above.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+
+	req := httptest.NewRequest("GET", "/", nil)
+	user := &auth.UserState{}
+	ucred := &ucrednet{uid: 0, pid: 100}
```
Please fill the socket information in.

zyga commented on daemon/access_test.go:
```diff
@@ -0,0 +1,186 @@
+	c.Check(ac.checkAccess(nil, ucred, user), Equals, accessForbidden)
+
+	// Non-root users are forbidden, even with macaroon auth
+	ucred = &ucrednet{uid: 42, pid: 100}
```
Please fill in the socket both here and in the test below.

zyga commented on daemon/api.go:
```diff
@@ -116,124 +116,136 @@ var api = []*Command{
-		UserOK: true,
-		GET:    getChanges,
+		Path:       "/v2/changes",
+		ReadAccess: openAccess{},
```
I have a small conceptual problem with this change. Before it was clear that it had to be a local access that was allowed. Now we call it `openAccess` and that's effectively the same but once we grow any kind of remote access existing distinction will be lost. It also bothers me somewhat that `open` implies only the classic world and secretly (though that's not something I disagree with) excludes access from snaps. 

How about we call `openAccess` `localSystemAccess` instead. This way we clearly indicate this is for local interactions only, we can add `remoteSystemAccess` later and we already have `snapAccess` that is both local (by definition) and non-system.

jhenstridge replied:
I chose the names to match what is used in the REST API documentation, which I think holds some value.  If we are going to change this, it should be something that would be understandable in the docs.

We can always rename things when TCP access to the API is a thing.

pedronis replied:
I don't think using the term "local" in the access name will help at this point. It's very likely also that the first form of remote access will be via the store, not accessing the snapd own socket over http. Indeed so far we tend to use "remote" to mean the store.

zyga commented on daemon/api.go:
```diff
@@ -2146,15 +2158,15 @@ func runSnapctl(c *Command, r *http.Request, user *auth.UserState) Response {
-	_, uid, _, err := runSnapctlUcrednetGet(r.RemoteAddr)
+	ucred, err := runSnapctlUcrednetGet(r.RemoteAddr)
 	if err != nil {
 		return Forbidden("cannot get remote user: %s", err)
```
Nitpick, perhaps this should say `requesting user` instead of `remote` as we may eventually have authenticated remote access over non-local sockets.

zyga reviewed (dismissed):
I reviewed the key changes to ucred and tests, and all of api.go. I'll park this review now and request changes for at least:

- explicit, up-front test for `ucred==nil` in each access check method.
- documentation for `openAccess`
- consideration for renaming `openAccess` to something else, as I've outlined in the inline review.

jhenstridge commented on tests/main/snapctl/task.yaml:
```diff
@@ -44,7 +44,7 @@ execute: |
-    if ! printf 'GET /v2/snaps HTTP/1.0\r\n\r\n' | nc -U -w 1 /run/snapd-snap.socket | grep '401 Unauthorized'; then
-        echo "Expected snapd API to be unauthorized on the snap socket"
+    if ! printf 'GET /v2/snaps HTTP/1.0\r\n\r\n' | nc -U -w 1 /run/snapd-snap.socket | grep '403 Forbidden'; then
+        echo "Expected snapd API to be forbidden on the snap socket"
```
With the existing access check code, it looks like you would be able to access the full API via `snapd-snap.socket` if you provide a macaroon.  However, it would not be possible to obtain a macaroon from `/v2/login` from that socket.

It's not clear that this is desirable behaviour, even if some snap has managed it somehow (using `personal-files` to access `~/.snap`? somehow finding one via `password-manager-service`?).

pedronis commented on daemon/api.go:
```diff
@@ -116,124 +116,136 @@ var api = []*Command{
-		GuestOK: true,
-		GET:     sysInfo,
+		Path:       "/v2/system-info",
+		ReadAccess: openAccess{},
```
I think the *Access should go after the GET/POST fields, like in the definition of the struct

pedronis commented on daemon/access.go:
```diff
@@ -0,0 +1,164 @@
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred == nil {
+		return accessForbidden
```
this is not tested.  There's  a bit of pointless variability btw on how the various checkAccess checks the same things. In n general all them except snapAccess do the same check as openAccess, so maybe the latter should also exist as a helper function that is called first by all of them.

pedronis commented on daemon/access.go:
```diff
@@ -0,0 +1,164 @@
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+// openAccess allows requests without authentication, provided they
```
There is some argument to made to call this "unrestricted" instead of "open", and then change the docs (which also atm aren't clear about the access restrictions on /v2/snapctl either) to make clear the role of the unix socket in the access. It seems something though to do in a self-contained follow up.

pedronis commented on daemon/access.go:
```diff
@@ -0,0 +1,163 @@
+	if ucred.uid == 0 {
+		return accessOK
+	}
+	return accessForbidden
```
this is an interesting case because while a token doesn't help switching to sudo would, basically it's uncelar that distinguishing only between Forbidden vs Authorize is enough to produce a good error above

jhenstridge replied:
The HTTP error codes are a bit misnamed.  The "401 Unauthorized" error code is described as meaning authentication is required:

https://tools.ietf.org/html/rfc2616#section-10.4.2

More specifically, the status code should be accompanied by a `WWW-Authenticate` header explaining how the client can construct an appropriate `Authorization` header to authenticate a future request.

In contrast, "403 Forbidden" means that the server is refusing to carry out the request and that an `Authorization` header won't help.  In any other system, this would be called "Unauthorized".

For endpoints that are relying on peer credentials, we've authenticated the client and decided that there is nothing that particular client could do to have the request accepted.  I agree that we still want to include a hint in 403 responses that a request by root will be handled.

pedronis commented on daemon/response.go:
```diff
@@ -192,7 +192,8 @@ func makeErrorResponder(status int) errorResponder {
-		if status == 401 {
+		// FIXME: forbidden should use a different kind.
+		if status == 401 || status == 403 {
 			res.Kind = client.ErrorKindLoginRequired
```
so at this point we actually would really like to know if we need to be root as well or not, it would be good to have a ErrorKindRootRequired

jhenstridge replied:
That sounds like a good idea.  Would it be acceptable to do this as a follow up?

pedronis replied:
yes, a follow up is ok, not sure this one will land before the follow up but otoh this is big enough that a follow up is a good idea anyway

pedronis commented on daemon/access.go:
```diff
@@ -0,0 +1,163 @@
+	return accessOK
+}
+
+// openAccess allows requests without authentication, provided they
```
I'm still leaning toward preferring unrestrictedAccess for this, but it can be a follow up, it would need changes to the docs as well

pedronis reviewed (commented):
this looks good, we have to figure out what to do with the error kinds though

pedronis reviewed (commented):
@jhenstridge I'm fundamentally happy with this, but given it needs 2 reviews, maybe it would go faster if the changes to ucred and simply the addition of access.go access_test.go would be done as two separate prereq PRs.  Then this would remain after those land as the changes to daemon.go and Commands and tests.

jhenstridge commented:
I've split out the ucrednet changes into PR #10126, and the access checkers into PR #10127.

jhenstridge commented:
The two related PRs have now been merged, so there should now be a cleaner diff to review only covering integration of the access checkers into the daemon.

pedronis commented on daemon/api_console_conf.go:
```diff
@@ -31,8 +31,9 @@ import (
-		POST: consoleConfStartRoutine,
+		Path:        "/v2/internal/console-conf-start",
+		POST:        consoleConfStartRoutine,
+		WriteAccess: authenticatedAccess{},
```
this probably could be just rootAccess,  cc @anonymouse64

anonymouse64 replied:
mmm yes /v2/internal/console-conf-start is only used by console-conf which runs as root when the user is driving it, so I think this is safe to always be root only and probably makes more sense to only allow it as a root only thing given that it is an internal endpoint

pedronis commented on daemon/api_model.go:
```diff
@@ -32,15 +32,16 @@ import (
+		POST:        postModel,
+		GET:         getModel,
+		ReadAccess:  openAccess{},
+		WriteAccess: authenticatedAccess{},
```
it might be a good idea to make this rootAccess actually

pedronis reviewed (commented):
this looks correct, I double checked in particular that all openAccess were UserOK or GuestOK before.

It can be landed once it gets 2nds revires and the follow up #10122  (which I need to rebase) is also approved for landing.

pedronis reviewed (approved):
I have rebased the follow-up #10122

bboozzoo commented on daemon/daemon.go:
```diff
@@ -97,91 +97,12 @@ type Command struct {
-	}
+	// Access control.
+	ReadAccess  accessChecker
+	WriteAccess accessChecker
```
An idea for a test. I think it would be useful to have a test similar to TestAddRoutes that walks all registered API commands and checks that WriteAccess is not nil iff PUT || POST are set, perhaps even verify that it is not an instance of openAccess.

jhenstridge replied:
We've essentially got that check here in `daemon_test.go`:

https://github.com/snapcore/snapd/blob/10f22a2a230e711042c61bd8d536e29fb56ed330/daemon/daemon_test.go#L343-L350

It doesn't specifically check for WriteAccess being set to openAccess, but it ensures that any command with a GET method sets ReadAccess, and any command with PUT or POST methods sets WriteAccess.

bboozzoo reviewed (approved):
LGTM and a nice improvement. Thank you.

pedronis commented:
with #10122 having two +1 now this is unblocked in principle

pedronis commented:
I'm going to merge master into this again and fix any divergences.
